File: 0002/AddTwoNumbers.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* addTwoNumbers(struct ListNode* l1, struct ListNode* l2) {
    struct ListNode *resultNode, *currentNode;
    int carry = 0;

    // Create an initial node
    resultNode = (struct ListNode *) malloc(sizeof(struct ListNode));
    resultNode->val = 0;
    resultNode->next = NULL;

    currentNode = resultNode;

    while (l1 != NULL || l2 != NULL) {
        int n1 = (l1 != NULL) ? l1->val : 0;
        int n2 = (l2 != NULL) ? l2->val : 0;
        int sum = n1 + n2 + carry;
        carry = sum / 10;

        currentNode->val = sum % 10;

        if (l1 != NULL) {
            l1 = l1->next;
        }

        if (l2 != NULL) {
            l2 = l2->next;
        }

        if (l1 != NULL || l2 != NULL) {
            // Create a next node
            struct ListNode *nextNode = (struct ListNode *) malloc(sizeof(struct ListNode));
            nextNode->val = 0;
            nextNode->next = NULL;
            currentNode->next = nextNode;
            currentNode = nextNode;
        }
    }

    if (carry > 0) {
        struct ListNode *nextNode = (struct ListNode *) malloc(sizeof(struct ListNode));
        nextNode->val = carry;
        nextNode->next = NULL;
        currentNode->next = nextNode;
        currentNode = nextNode;
    }

    return resultNode;
}

File: 0002/AddTwoNumbers.swift
public class ListNode {
    public var val: Int
    public var next: ListNode?
    public init() { self.val = 0; self.next = nil; }
    public init(_ val: Int) { self.val = val; self.next = nil; }
    public init(_ val: Int, _ next: ListNode?) { self.val = val; self.next = next; }
}

class Solution {
    func addTwoNumbers(_ l1: ListNode?, _ l2: ListNode?) -> ListNode? {
        var addedValue = 0
        var position = 0
        var l1CurrentNode: ListNode? = l1
        var l2CurrentNode: ListNode? = l2
        var resultListNode = ListNode(0)
        var carryNode: ListNode?
        var currentNode = resultListNode

        while l1CurrentNode != nil || l2CurrentNode != nil {
            addedValue += (l1CurrentNode != nil) ? l1CurrentNode!.val : 0
            addedValue += (l2CurrentNode != nil) ? l2CurrentNode!.val : 0
            addedValue += (carryNode != nil) ? carryNode!.val : 0

            var newNode: ListNode?
            if position == 0 {
                resultListNode.val = addedValue
                newNode = resultListNode
            } else {
                newNode = ListNode(addedValue)
                currentNode.next = newNode
                currentNode = newNode!
            }
            carryNode = checkCarrying(newNode!)

            l1CurrentNode = l1CurrentNode?.next ?? nil
            l2CurrentNode = l2CurrentNode?.next ?? nil
            addedValue = 0
            position += 1
        }

        if carryNode != nil {
            currentNode.next = carryNode
            currentNode = carryNode!
            position += 1
        }

        return resultListNode
    }

    func checkCarrying(_ node: ListNode) -> ListNode? {
        var carryNode: ListNode? = nil

        if node.val / 10 > 0 {
            carryNode = ListNode(node.val / 10)
            var remainder = node.val % 10
            node.val = remainder
        }

        return carryNode
    }
}
